craft_cli/messages.py

```diff
@@ -252,6 +265,33 @@ def trace(self, text: str) -> None:
         stream = sys.stderr if self.mode == EmitterMode.TRACE else None
         self.printer.show(stream, text, use_timestamp=True)  # type: ignore
 
+    @_init_guard
+    def progress(self, text: str) -> None:
+        """Progress information for a multi-step command.
+
+        This is normally to present several separated text messages.
+
+        These messages will be truncated to the terminal's width, and overwritten by the next
+        line (unless verbose/trace mode).
+        """
+        if self.mode == EmitterMode.QUIET:
+            # will not be shown in the screen (always logged to the file)
+            stream = None
+            use_timestamp = False
+            ephemeral = True
+        elif self.mode == EmitterMode.NORMAL:
+            # show the indicated message to stderr (ephemeral) and log it
+            stream = sys.stderr
```

Comment:
Why are progress messages sent to stderr instead of stdout? Wouldn't that result in unexpected output patterns e.g. when someone runs the application redirecting or piping stdout (and expecting all messages to go to the redirected destination)?

Reply:
The idea is: **Everything** is sent to stderr, except for the "expected command output" (see [the table](https://docs.google.com/document/d/1Pe-0ED6db53SmrUGIAgVMzxeCOGJQwsZJWf7jzFSagQ)) so if you want to run `charmcraft` from other process and capture the result, you just send stderr to /dev/null and whatever you got is just the result of the command.